File: data_structures/python/SinglyLinearLL.py
class Node:                                      #class
    def __init__(self, data):                    #constructor
        self.data = data                         #instance variable or attributes
        self.next = None                         #instance variable

class SinglyLinearLL:                           #class
    def __init__(self):                           #constructor
        self.head = None                          #instance variable

    def InsertFirst(self, data):                  #instance method
        new_node = Node(data)
        if self.head == None:#LL is empty
            self.head = new_node
            return
        
        else:#LL has atleast one Node
            new_node.next = self.head
            self.head = new_node

    def InsertLast(self, data):
        new_node = Node(data)
        if self.head == None:#LL is empty
            self.head = new_node
            return

        temp = self.head
        while temp.next:
            temp = temp.next
        temp.next = new_node

    def DeleteFirst(self):
        if self.head == None:#LL is empty
            print("Linked-List is empty.")
            return
        
        if self.head.next == None:#LL has only one node
            self.head = None

        else:#LL has atleast one node
            self.head = self.head.next

    def DeleteLast(self):
        if self.head == None:
            print("Linked-List is empty.")
            return
        
        if self.head.next == None:
            self.head = None
            return

        temp = self.head
        while temp.next.next:
            temp = temp.next
        temp.next = None

    def Display(self):
        if self.head == None:
            print("List is empty.")
            return

        temp = self.head
        while temp != None:
            print(f"| {temp.data} | -> ", end="")
            temp = temp.next
        print("None")

    def Count(self):
        iCount = 0
        temp = self.head

        while temp != None:
            iCount += 1
            temp = temp.next
        return iCount


    def InsertAtPos(self,data,pos):
        new_node = Node(data)

        temp = self.head

        iCount = self.Count()

        if(pos < 1 or pos > iCount + 1):
            print("Invalid position")

        elif(pos == 1):
            self.InsertFirst(data)

        elif(pos == iCount + 1):
            self.InsertLast(data)

        else:
            for i in range(1,pos - 1):
                temp = temp.next

            new_node.next = temp.next
            temp.next = new_node


    def DeleteAtPos(self,pos):
        temp = self.head

        iCount = self.Count()

        if(pos < 1 or pos > iCount):
            print("Invalid position")

        elif(pos == 1):
            self.DeleteFirst()

        elif(pos == iCount):
            self.DeleteLast()

        else:
            for i in range(1,pos - 1):
                temp = temp.next
            target = temp.next
            temp.next = target.next

def main():

    sobj = SinglyLinearLL()

    sobj.InsertFirst(51)
    sobj.InsertFirst(21)
    sobj.InsertFirst(11)

    sobj.InsertLast(101)
    sobj.InsertLast(111)
    sobj.InsertLast(121)

    sobj.Display()  
    iRet = sobj.Count()
    print("The total number of nodes in linked list are: ", iRet) 

    sobj.InsertAtPos(105,5)

    sobj.Display()  
    iRet = sobj.Count()
    print("The total number of nodes in linked list are: ", iRet) 

    sobj.DeleteAtPos(5)
    
    sobj.Display()  
    iRet = sobj.Count()
    print("The total number of nodes in linked list are: ", iRet) 

if __name__ == "__main__":
    main()

File: data_structures/C++/SinglyLinearLL.cpp
#include<iostream>
using namespace std;

struct node
{
    int data;
    struct node *next;
};

typedef struct node NODE;
typedef struct node * PNODE;
typedef struct node ** PPNODE;

class SinglyLinearLL
{
    private:
    PNODE first;
    int iCount;

    public:
    SinglyLinearLL()
    {
        cout<<"Inside Constructor of SinglyLinearLL\n";
        this->first = NULL;
        this->iCount = 0;
    }
    void InsertFirst(int no)
    {
        PNODE newn = NULL;
        newn = new NODE;
        newn->data = no;
        newn->next = NULL;

        if(first == NULL)// LL is empty
        {
            first = newn;
        }
        else// LL contains atleast one node
        {
            newn->next = first;
            first = newn;
        }
        iCount++;

    }
    void InsertLast(int no)
    {
        PNODE newn = NULL;
        newn = new NODE;
        newn->data = no;
        newn->next = NULL;

        if(first == NULL)// LL is empty
        {
            first = newn;
        }
        else// LL contains atleast one node
        {
           PNODE temp = NULL;
           temp = first;

           while(temp->next != NULL)
           {
                temp = temp->next;
           }
           temp->next = newn;
        }
        iCount++;
    }
    void DeleteFirst()
    {
        if(first == NULL)// LL is empty
        {
            cout<<"Linked List is empty\n";
            return;
        }
        if (first->next == NULL)// LL has only one node
        {
            delete first;
            first = NULL;
        }
        else
        {   
            PNODE temp = NULL;
            temp = first;
            first = first->next;
            delete temp;

            
        }
        iCount--;

    }
    void DeleteLast()
    {
        if(first == NULL)// LL is empty
        {
            cout<<"Linked List is empty\n";
            return;
        }
        if (first->next == NULL)// LL has only one node
        {
            delete first;
            first = NULL;
        }
        else
        {
            PNODE temp = NULL;
            temp = first;
            while(temp->next->next != NULL)
            {
                temp = temp->next;
            }
            delete temp->next;
            temp->next = NULL;

            
        }
        iCount--;
    }
    void Display()
    {
        if(first == NULL)
        {
            cout<<"Linked List is empty\n";
            return;
        }

        PNODE temp = NULL;
        temp = first;
        while(temp != NULL)
        {
            cout<<"| "<<temp->data<<" |->";
            temp = temp->next;
        }
        cout<<"NULL\n";

    }
    int Count()
    {
        return iCount;
    }
    void InsertAtPos(int no,int pos)
    {   
        iCount = Count();

        if(pos < 1 || pos > iCount + 1)
        {
            cout<<"Invalid Position\n";
            return;
        }

        if(pos == 1)
        {
            InsertFirst(no);
        }
        else if(pos == iCount + 1)
        {
            InsertLast(no);
        }
        else
        {
            PNODE newn = new NODE;
            newn->data = no;
            newn->next = NULL;

            PNODE temp = NULL;
            temp = first;
            int iCnt = 0;

            for(iCnt = 1;iCnt < pos - 1;iCnt++)
            {
                temp = temp->next;
            }
            newn->next = temp->next;
            temp->next = newn;

            iCount++;
        }

    }
    void DeleteAtPos(int pos)
    {
        iCount = Count();

        if(pos < 1 || pos > iCount)
        {
            cout<<"Invalid Position\n";
            return;
        }

        if(pos == 1)
        {
            DeleteFirst();
        }
        else if(pos == iCount)
        {
            DeleteLast();
        }
        else
        {
            PNODE temp = NULL;
            temp = first;
            PNODE target = NULL;
            int iCnt = 0;

            for(iCnt = 1;iCnt < pos - 1;iCnt++)
            {
                temp = temp->next;
            }
            target = temp->next;
            temp->next = target->next;
            delete target;

            iCount--;
        }
    }
};

int main()
{
    SinglyLinearLL sobj;
    int iRet = 0;

    sobj.InsertFirst(51);
    sobj.InsertFirst(21);
    sobj.InsertFirst(11);
    
    sobj.InsertLast(101);
    sobj.InsertLast(111);
    sobj.InsertLast(121);

    sobj.Display();
    
    iRet = sobj.Count();

    cout<<"Number of elements are : "<<iRet<<"\n";
    
    sobj.InsertAtPos(105,5);

    sobj.Display();
    
    iRet = sobj.Count();

    cout<<"Number of elements are : "<<iRet<<"\n";

    sobj.DeleteAtPos(5);

    sobj.Display();
    
    iRet = sobj.Count();

    cout<<"Number of elements are : "<<iRet<<"\n";

    return 0;
}